File: evaluation/results/scenario_2/stan/llama3.1:8b/run_03/extracted_code.R
# Install necessary packages
install.packages("cmdstanr")
install.packages("dplyr")

# Load necessary libraries
library(cmdstanr)
library(dplyr)

# Read data
cases <- read.csv("cases.csv", colClasses = c("Date", "integer"))

# Prepare data for Stan
cases$day_of_week <- factor(cases$day_of_week, levels = 1:7, ordered = TRUE)
cases$date <- as.Date(cases$date)

# Define the model in Stan

model_code <- "
data {
  int<lower=0> T; // number of time points
  vector[T] cases; // reported cases at each time point
  vector[T] day_of_week; // day-of-week at each time point (1-7)
  real<lower=0, upper=1> ascertainment[T]; // proportion of infections that are reported at each time point
}

parameters {
  real<lower=0> R; // average reproduction number
  vector[T] gamma; // delay between infection and reporting at each time point (days)
  real<lower=0, upper=1> phi_day_of_week[7]; // day-of-week effect on reported cases (7 levels)
  real<lower=0, upper=1> alpha; // overdispersion parameter
}

model {
  // Model for day of week effects
  for (t in 1:T) {
    gamma[t] ~ normal(3, 2); // prior for delay between infection and reporting
    phi_day_of_week[day_of_week[t]] ~ beta(1, 1); // prior for day-of-week effect on reported cases
  }

  // Model for observed cases (renewal equation)
  for (t in 1:T) {
    R ~ normal(log(mean(gamma)) + 0.5 * (log(sum(cases[gamma < mean(gamma)])) - log(T)), sqrt(2^2));
    target += poisson_log_lik(cases[t], 
                              R * sum(phi_day_of_week[day_of_week[t]] .* cases[(gamma < T) & (gamma >= t)]) .* ascertainment[t]);
  }
  
  // Model for overdispersion
  alpha ~ beta(1, 1);
}
"

# Fit the model

fit <- cmdstanr::sample(
  file = "reproduction_number.stan", 
  data = list(T = nrow(cases), cases = cases$cases, day_of_week = as.numeric(factor(cases$day_of_week)), ascertainment = rep(1, nrow(cases))),
  chains = 4, iter_warmup = 1000, iter_sampling = 2000
)

# Extract results

Rt_estimates <- fit$summary("R", probs = c(0.025, 0.975))
day_of_week_effects <- fit$summary("phi_day_of_week", probs = c(0.025, 0.975))
ascertainment_estimates <- fit$summary("ascertainment", probs = c(0.025, 0.975))

# Plot the results

library(ggplot2)

ggplot(cases, aes(x = date)) +
  geom_line(aes(y = Rt_estimates[, 1]), color = "red") +
  geom_point(data = cases, aes(y = Rt_estimates[, 2]), color = "blue") +
  labs(title = "Estimated Time-Varying Reproduction Number (Rt)", x = "", y = "") +
  theme_classic()

ggplot(cases, aes(x = day_of_week)) +
  geom_bar(aes(fill = factor(day_of_week)), position = "dodge", color = "black") +
  labs(title = "Day-of-Week Effects on Reported Cases", fill = "Day of Week")

ggplot(cases, aes(x = date)) +
  geom_line(aes(y = ascertainment_estimates[, 1]), color = "red") +
  geom_point(data = cases, aes(y = ascertainment_estimates[, 2]), color = "blue") +
  labs(title = "Estimated Time-Varying Ascertainment", x = "", y = "") +
  theme_classic()

File: evaluation/results/scenario_2/pymc/llama3.1:8b/run_01/extracted_code.py
import pandas as pd
import numpy as np
import pymc as pm
import matplotlib.pyplot as plt

# Load data
df = pd.read_csv('cases.csv')
df['date'] = pd.to_datetime(df['date'])

# Define constants
generation_interval = 5.0  # Average generation interval in days

# Fit model
with pm.Model() as model:
    # Day-of-week effects (multiplicative)
    day_of_week Effects = pm.Normal('day_of_week_effects', mu=1, sigma=0.2,
                                    shape=(7,))
    
    # Time-varying ascertainment (smoothly varying proportion of infections reported)
    ascertainment = pm.Uniform('ascertainment', lower=0, upper=1,
                               transformed=pm.distributions.transforms.Log,
                               shape=(len(df),))
    
    # Infection dynamics
    I_past = df['cases'].shift(generation_interval).values  # Lagged cases by generation interval
    
    # Renewal equation with day-of-week effects and ascertainment
    Rt = pm.Deterministic('Rt', (day_of_week_effects * I_past) * ascertainment)
    
    # Observation model (Poisson with overdispersion)
    observed_cases = pm.Poisson('observed_cases', mu=Rt,
                                observed=df['cases'],
                                shape=(len(df),),
                                tol=1e-6)
    
# Run MCMC
with model:
    idata = pm.sample(2000, tune=1000)

# Extract estimates
Rt_estimates = idata.posterior['Rt'].mean(axis=0).values
day_of_week_effect_estimates = idata.posterior['day_of_week_effects'].mean(axis=0).values
ascertainment_estimates = idata.posterior['ascertainment'].mean(axis=0).values

# Plot results
plt.figure(figsize=(10, 6))
plt.plot(Rt_estimates)
plt.title('Estimated Time-Varying Reproduction Number (Rt)')
plt.xlabel('Time')
plt.ylabel('Rt estimate')
plt.show()

plt.figure(figsize=(10, 6))
plt.bar(range(1, 8), day_of_week_effect_estimates)
plt.title('Day-of-Week Effect Estimates')
plt.xlabel('Day of Week')
plt.ylabel('Effect Estimate')
plt.show()

plt.figure(figsize=(10, 6))
plt.plot(ascertainment_estimates)
plt.title('Time-Varying Ascertainment Estimates')
plt.xlabel('Time')
plt.ylabel('Ascertainment estimate')
plt.show()

